src/pudl/analysis/allocate_net_gen.py

```diff
@@ -324,183 +509,437 @@ def stack_generators(
+    )
+    gf_pm_fuel_summed = (
+        gf.groupby(by=IDX_PM_ESC)
+        .sum(min_count=1)[DATA_COLUMNS]
+        .add_suffix("_gf_tbl")
+        .reset_index()
+    )
 
     gen_assoc = (
-        pd.merge(stack_gens, gen, on=IDX_GENS, how="outer")
-        .pipe(remove_retired_generators)
+        pd.merge(
+            stack_gens,
+            gen.rename(columns={"net_generation_mwh": "net_generation_mwh_g_tbl"}),
+            on=IDX_GENS,
+            how="outer",
+        )
         .merge(
-            gf.groupby(by=IDX_PM_ESC, as_index=False).sum(min_count=1),
+            gf_pm_fuel_summed,
             on=IDX_PM_ESC,
-            suffixes=("_g_tbl", "_gf_tbl"),
             how="outer",
+            validate="m:1",
+            indicator=True,  # used in _allocate_unassociated_records to find unassocited
         )
-    )
-
-    gen_assoc = (
-        pd.merge(
-            gen_assoc,
-            gen_assoc.groupby(by=IDX_ESC)[["capacity_mw", "net_generation_mwh_g_tbl"]]
-            .sum(min_count=1)
-            .add_suffix("_fuel")
-            .reset_index(),
-            on=IDX_ESC,
+        .pipe(remove_inactive_generators)
+        .pipe(
+            _allocate_unassociated_records,
+            idx_cols=IDX_PM_ESC,
+            col_w_unexpected_codes="prime_mover_code",
+            data_columns=[f"{col}_gf_tbl" for col in DATA_COLUMNS],
         )
-        .pipe(apply_pudl_dtypes, "eia")
-        .pipe(_associate_unconnected_records)
+        .drop(columns=["_merge"])  # drop do we can do this again in the bf_summed merge
+        .merge(
+            bf_summed, on=IDX_GENS_PM_ESC, how="outer", validate="m:1", indicator=True
+        )
+        .pipe(
+            _allocate_unassociated_records,
+            idx_cols=IDX_GENS_PM_ESC,
+            col_w_unexpected_codes="energy_source_code",
+            data_columns=["fuel_consumed_mmbtu_bf_tbl"],
+        )
+        .drop(columns=["_merge"])
     )
+
+    # replace zeros with small number to avoid div by zero errors when calculating allocation fraction
+    data_columns = [
+        "fuel_consumed_mmbtu_gf_tbl",
+        "fuel_consumed_for_electricity_mmbtu_gf_tbl",
+        "net_generation_mwh_gf_tbl",
+        "fuel_consumed_mmbtu_bf_tbl",
+    ]
+    gen_assoc[data_columns] = gen_assoc[data_columns].replace(0, MISSING_SENTINEL)
+
+    # calculate the total capacity in every fuel group
+    gen_assoc = pd.merge(
+        gen_assoc,
+        gen_assoc.groupby(by=IDX_ESC)[["capacity_mw", "net_generation_mwh_g_tbl"]]
+        .sum(min_count=1)
+        .add_suffix("_fuel")
+        .reset_index(),
+        on=IDX_ESC,
+        how="outer",
+    ).pipe(apply_pudl_dtypes, "eia")
     return gen_assoc
 
 
-def remove_retired_generators(gen_assoc):
+def remove_inactive_generators(gen_assoc: pd.DataFrame) -> pd.DataFrame:
     """Remove the retired generators.
 
     We don't want to associate net generation to generators that are retired
```

Comment:
Should this be "associate ... with" or "allocate ... to?" and should it mention both net generation and fuel consumption?

Reply:
this gets employed inside of the association step. but i can add "and later allocate"

and fair.